File: android/app/src/main/java/com/enctxt/core/privacy/ProtectionStylePreference.kt
package com.enctxt.core.privacy

import android.content.Context
import android.content.SharedPreferences

/**
 * Local Protection Style Preference Storage
 *
 * Persists the user's chosen rendering mode (Classic/Illusion/Pattern) locally on-device.
 *
 * SECURITY & PRIVACY NOTICE:
 * - This preference is strictly LOCAL ONLY.
 * - It is a UI/rendering choice, not sensitive data (unlike the gesture sequence, which uses
 *   [com.enctxt.core.gesture.EncryptedGestureStorage]) — plain [SharedPreferences] is sufficient.
 *   It is never sent to the backend and does not need to be synchronized across devices.
 */
interface ProtectionStylePreference {
    fun getMode(userId: String): ProtectedRenderMode
    fun setMode(userId: String, mode: ProtectedRenderMode): Boolean
}

class SharedPrefsProtectionStylePreference(context: Context) : ProtectionStylePreference {

    private val appContext = context.applicationContext

    private val prefs: SharedPreferences by lazy {
        appContext.getSharedPreferences(PREFS_FILE_NAME, Context.MODE_PRIVATE)
    }

    override fun getMode(userId: String): ProtectedRenderMode {
        if (userId.isBlank()) return ProtectedRenderMode.HOMOGLYPH

        val stored = try {
            prefs.getString(keyFor(userId), null)
        } catch (_: Exception) {
            null
        }

        return stored?.let { name ->
            ProtectedRenderMode.entries.find { it.name == name }
        } ?: ProtectedRenderMode.HOMOGLYPH
    }

    override fun setMode(userId: String, mode: ProtectedRenderMode): Boolean {
        if (userId.isBlank()) return false

        return try {
            prefs.edit().putString(keyFor(userId), mode.name).apply()
            true
        } catch (_: Exception) {
            false
        }
    }

    private fun keyFor(userId: String) = "${KEY_PREFIX}$userId"

    companion object {
        private const val PREFS_FILE_NAME = "protected_text_prefs"
        private const val KEY_PREFIX = "protection_style_"
    }
}

File: client/src/utils/protectedText/protectionStylePreference.ts
/**
 * Local Protection Style Preference Storage
 *
 * Persists the user's chosen rendering mode (Classic/Illusion/Pattern) locally in
 * `localStorage`, mirroring the pattern already used by `gestureStorage.ts`.
 *
 * SECURITY & PRIVACY NOTICE:
 * - This preference is strictly LOCAL ONLY.
 * - It is a UI/rendering choice, never sent to the backend, never stored in cookies, and never
 *   transmitted in WebSocket frames — it does not need to be synchronized across devices.
 */

import type { ProtectionMode } from './protectedTextEngine';

export const CURRENT_PREFERENCE_VERSION = 1;
const STORAGE_PREFIX = 'enctxt_protection_style_';

const VALID_MODES: ProtectionMode[] = ['HOMOGLYPH', 'ILLUSION', 'PATTERN'];

interface StoredProtectionStyle {
  version: number;
  mode: ProtectionMode;
}

function getStorageKey(userId: string): string {
  return `${STORAGE_PREFIX}${userId}`;
}

/**
 * Reads the user's stored protection style preference. Falls back to 'HOMOGLYPH' (the existing
 * "Classic" behavior) if nothing is stored or the stored value is invalid/corrupted.
 */
export function getProtectionStyle(userId: string): ProtectionMode {
  if (!userId) return 'HOMOGLYPH';

  try {
    const raw = localStorage.getItem(getStorageKey(userId));
    if (!raw) return 'HOMOGLYPH';

    const parsed = JSON.parse(raw) as StoredProtectionStyle;
    if (parsed && VALID_MODES.includes(parsed.mode)) {
      return parsed.mode;
    }
    return 'HOMOGLYPH';
  } catch {
    return 'HOMOGLYPH';
  }
}

/**
 * Saves the user's protection style preference locally.
 */
export function setProtectionStyle(userId: string, mode: ProtectionMode): boolean {
  if (!userId || !VALID_MODES.includes(mode)) return false;

  try {
    const data: StoredProtectionStyle = { version: CURRENT_PREFERENCE_VERSION, mode };
    localStorage.setItem(getStorageKey(userId), JSON.stringify(data));
    return true;
  } catch (error) {
    console.error('Failed to save protection style preference to local storage:', error);
    return false;
  }
}
